nixos/modules/services/networking/yggdrasil.nix

```diff
@@ -0,0 +1,119 @@
+          network each time.
+
+          If this option is set, <option>config</option> will be ignored.
+        '';
+      };
+
+      config = mkOption {
+        type = str;
+        default = "";
+        description = ''
+          JSON or HJSON configuration for yggdrasil.
+
+          Warning: this is stored in the WORLD-READABLE Nix store!
+          Therefore, it is not appropriate for private keys.  If you
+          provide a partial configuration and do not specify the keys,
+          yggdrasil will generate a new set each time the service is
+          started, creating a random IPv6 address on the yggdrasil
+          network each time.
+
+          If you wish to specify the keys, use
+          <option>configFile</option>, which takes precedence over
+          <option>config</option>.
+        '';
+      };
+
+      port = mkOption {
+        type = nullOr int;
+        default = null;
+        example = 54321;
+           description = ''
+          Firewall port to open in order to permit incoming peerings.
+          This should match the port specified in <code>Listen</code>
+          in the yggdrasil configuration.  It is not necessary to open
+          a port in order to make outgoing peerings.
+        '';
+      };
+
+      openMulticastPort = mkOption {
+        type = bool;
+        default = true;
+        description = ''
+          Whether to open the UDP port used for multicast peer discovery
+          on local networks.
+        '';
+      };
+
+      package = mkOption {
+        type = package;
+        default = pkgs.yggdrasil;
+        defaultText = "pkgs.yggdrasil";
+        description = "Yggdrasil package to use.";
+      };
+
+
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.yggdrasil.configFile = mkDefault (
+      if cfg.config != "" then
+        toString(pkgs.writeTextFile {
+          name = "yggdrasil-conf";
+          text = cfg.config;
+        })
+      else null);
+
+    systemd.services.yggdrasil = {
+      description = "Yggdrasil Network Service";
+      path = [ cfg.package ];
+      bindsTo = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        mkdir -p /var/lib/yggdrasil/
+        chmod 700 /var/lib/yggdrasil
+        chown -R root:root /var/lib/yggdrasil
+        '' + (if (cfg.configFile == null) then ''
+           ${cfg.package}/bin/yggdrasil -genconf > /var/lib/yggdrasil/yggdrasil.conf
+        '' else ''
+          ${cfg.package}/bin/yggdrasil -normaliseconf -useconffile \
+            ${cfg.configFile}  > /var/lib/yggdrasil/yggdrasil.conf
+        '');
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/yggdrasil -useconffile /var/lib/yggdrasil/yggdrasil.conf";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        Restart = "always";
+      };
+    };
+
+    networking.firewall.allowedTCPPorts = mkIf (cfg.port != null) [ cfg.port ];
```

Comment:
https://nixos.org/nixpkgs/manual/#reviewing-contributions-new-modules
`For example, enabling a module should not open firewall ports by default. `

Reply:
I don't understand what you're asking for here.  The default for `cfg.port` is `null` so a port only gets opened if you set it to something else.

Reply:
But why? maybe people want to use in on another port and require special firewall rules? And why only open it, if it's a non default port?

Reply:
Just as a convenience to keep everything related to yggdrasil configuration together, in the presumably more common case when special firewall rules are not needed.

Reply:
ok, but it's very hard to have a different port configured and not open the ports on all interfaces then. I would prefer it, if this would be left out of the module.

Reply:
I removed it.